File: Go/Day01.go
package main

import AH "./adventhelper"

func part1(v []int, t int) int {
	for r, i := range v {
		for _, j := range v[r:] {
			if i + j == t {
				return i * j
			}			
		}
	}

	return -1
}

func part2(v []int, t int) int {
	for r, i := range v {
		for s, j := range v[r:] {
			if (i + j > t) {
				continue;
			}
			for _, k := range v[s:] {
				if i + j + k == t {
					return i * j * k
				}
			}
		}		
	}

	return -1
}

func main() {
	ints, _ := AH.ReadIntFile("../input/input01.txt")

	AH.PrintSoln(1, part1(ints, 2020), part2(ints, 2020))

	return
}

File: Haskell/Day01.hs
import AdventHelper

sum_and_prod :: Integer -> [(Integer, Integer)] -> [(Integer, Integer)] -> [(Integer, Integer)]
sum_and_prod _ [x] _ = []
sum_and_prod t (x:xs) ys = q ++ (sum_and_prod t xs ys)
  where q = map (\(s',p') -> ((fst x)+s', (snd x)*p')) ys

main = do
  putStrLn "Day 1"
  f <- readFile "../input/input01.txt"
  let s = map(read) $ lines f :: [Integer]
  let p0 = zip s s
  let t = 2020

  let p1 = sum_and_prod t p0 p0
  printSoln 1 (snd $ head (dropWhile (\l -> fst l /= t) p1))
  let p2 = sum_and_prod t (filter (\(s,_) -> s < t) p1) p0
  printSoln 2 (snd $ head (dropWhile (\l -> fst l /= t) p2))
